resources/splunk_install.rb

```diff
@@ -0,0 +1,219 @@
+# Encoding: utf-8
+
+class SplunkInstall < ChefCompat::Resource
+  include PlatformHelpers, PathHelpers
+
+  resource_name :splunk_install
+
+  property :name, String, name_property: true, identity: true
+  property :package, [:splunk, :universal_forwarder], required: true
+  property :service_user, String, default: 'splunk'
+  property :service_password, String
+  property :version, String, required: true
+  property :build, String, required: true
+  property :source_url, String, default: 'https://download.splunk.com/products'
+  property :accept_license, [TrueClass, FalseClass], default: false
+
+  default_action :install
+
+  def after_created
+    # Validate Package
+    case name.downcase
+    when 'splunk' then package :splunk
+    when 'universal_forwarder' then package :universal_forwarder
+    else raise 'Package must be specified (:splunk or :universal_forwarder)'
+    end unless property_is_set? :package
+  end
+
+  ### Inherited Methods
+
+  # Must be overridden by a platform provider
+  def kernel_string
+    raise 'No implementation for current platform'
+  end
+
+  def package_name
+    case package
+    when :splunk then 'splunk'
+    when :universal_forwarder then 'splunkforwarder'
+    end
+  end
+
+  def install_dir
+    @install_dir ||= default_install_dirs[package][node['os'].to_sym]
+    raise "Unsupported Combination: #{package} + #{node['os']}" unless @install_dir
+    @install_dir
+  end
+
+  def package_url
+    @package_url ||=
+      case package
+      when :splunk then "#{source_url}/splunk/releases/#{version}/#{node['os']}/splunk"
+      when :universal_forwarder then "#{source_url}/universalforwarder/releases/#{version}/#{node['os']}/splunkforwarder"
+      end + "-#{version}-#{build}-#{kernel_string}"
+  end
+
+  def package_path
+    @package_path ||= Pathname.new(Chef::Config['file_cache_path']) + filename_from_url(package_url)
+  end
+
+  ### Inherited Actions
+
+  load_current_value do |desired|
+    package desired.package
+    version_file = version_pathname
+    current_value_does_not_exist! unless version_file.exist?
+
+    version_data = version_file.read
+    installed_version, installed_build = version_data.match(/VERSION=(\d(?:\.\d)+).+BUILD=([\w]+)/m).captures
+    version installed_version
+    build installed_build
+  end
+
+  action :install do
+    state = node.run_state['splunk_ingredient'] ||= {}
+    state['packages'] ||= {}
+
+    raise "Installation resource with name #{name} already exists!" if state['packages'][name]
```

Comment:
We should key by install_dir. I forgot I moved this check to install, and I wanted to keep the name so I could easily test uninstall.

We should probably store installations in run_state instead of packages, as that's what the splunk_install resource deals with. I'm curious now, though, if I could/should load existing installations into run_state at compile time.